src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/FileExtensionsAttributeAdapter.cs

```diff
@@ -0,0 +1,45 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.DataAnnotations;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Validation;
+using Microsoft.Extensions.Localization;
+
+namespace Microsoft.AspNetCore.Mvc.DataAnnotations.Internal
+{
+    public class FileExtensionsAttributeAdapter : AttributeAdapterBase<FileExtensionsAttribute>
+    {
+        public FileExtensionsAttributeAdapter(FileExtensionsAttribute attribute, IStringLocalizer stringLocalizer)
+            : base(attribute, stringLocalizer)
+        {
+        }
+
+        /// <inheritdoc />
+        public override void AddValidation(ClientModelValidationContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            MergeAttribute(context.Attributes, "data-val", "true");
+            MergeAttribute(context.Attributes, "data-val-fileextensions", GetErrorMessage(context));
+            MergeAttribute(context.Attributes, "data-val-fileextensions-extensions", Attribute.Extensions.ToLowerInvariant());
```

Comment:
Eh, I don't agree with the previous comments about normalization. When do we ever "normalize" file paths? What does it even mean to "normalize" a file paths? I think if you type "*.FOO" then you get "*.FOO", no?

Reply:
The code @kichalla has here isn't exactly normalizing a file path. It's about the `FileExtensionsAttribute.Extensions` property and how it's matched against the associated `string` property. `FileExtensionsAttribute` does that case-insensitively and keeps lower-cased `string`s around for that purpose.

See [here](https://github.com/dotnet/corefx/blob/15f9a31fd45f9a7d667765ef08ecfd1cefddde3a/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/FileExtensionsAttribute.cs#L35-L41) and [here](https://github.com/dotnet/corefx/blob/15f9a31fd45f9a7d667765ef08ecfd1cefddde3a/src/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/FileExtensionsAttribute.cs#L66) in `FileExtensionsAttribute`.

Reply:
Oh that's super weird that the attribute does that "normalization" (and it *is* normalization, it's just *incorrect* normalization).

But... I guess because the attribute already does this on the server (in a bizarre way), I guess the discussion moves over to https://github.com/aspnet/jquery-validation-unobtrusive/pull/70#discussion_r139814330, where we're discussing the client side, because we presumably need to make sure the client side matches whatever (odd) behavior the server has.